application/src/test/java/org/opentripplanner/street/model/edge/StreetEdgePropulsionCostTest.java

```diff
@@ -0,0 +1,348 @@
+import org.opentripplanner.routing.api.request.preference.VehicleRentalPreferences;
+import org.opentripplanner.routing.core.VehicleRoutingOptimizeType;
+import org.opentripplanner.service.vehiclerental.model.RentalVehicleType.PropulsionType;
+import org.opentripplanner.street.model.RentalFormFactor;
+import org.opentripplanner.street.model.StreetTraversalPermission;
+import org.opentripplanner.street.model.vertex.StreetVertex;
+import org.opentripplanner.street.search.request.StreetSearchRequest;
+import org.opentripplanner.street.search.state.StateEditor;
+
+/**
+ * Tests for propulsion-aware cost calculation in StreetEdge.
+ * <p>
+ * Verifies that different propulsion types (ELECTRIC, ELECTRIC_ASSIST, HUMAN) result in
+ * appropriate cost calculations, especially regarding slope effects:
+ * - ELECTRIC (e-scooters): Use flat distance (constant speed, motor does all work)
+ * - ELECTRIC_ASSIST (e-bikes): Reduced slope sensitivity (30% of human-powered effect)
+ * - HUMAN and others: Full slope effect
+ */
+class StreetEdgePropulsionCostTest {
+
+  private static final double DELTA = 0.00001;
+  private static final double SPEED = 6.0;
+  private static final double LENGTH = 650.0;
+
+  private final StreetVertex v0 = intersectionVertex(0.0, 0.0);
+  private final StreetVertex v1 = intersectionVertex(2.0, 2.0);
+
+  @Test
+  void rentalScooterCanTraverseEdge() {
+    StreetEdge edge = streetEdge(v0, v1, 100.0, StreetTraversalPermission.ALL);
+    var req = StreetSearchRequest.of().withMode(StreetMode.SCOOTER_RENTAL).build();
+    var editor = new StateEditor(v0, req);
+    editor.beginFloatingVehicleRenting(
+      RentalFormFactor.SCOOTER,
+      PropulsionType.ELECTRIC,
+      "network",
+      false
+    );
+    var state = editor.makeState();
+
+    var result = edge.traverse(state);
+    assertNotNull(result);
+    assertEquals(1, result.length);
+    assertEquals(PropulsionType.ELECTRIC, result[0].rentalVehiclePropulsionType());
+  }
+
+  /**
+   * Tests that electric scooters ignore slope for time calculation.
+   * On a hilly street with elevation profile, an electric scooter should use flat distance
+   * for its time estimate since the motor maintains constant speed.
+   * <p>
+   * Uses TRIANGLE optimization with time=1 to isolate the time/speed component,
+   * similar to the approach in StreetEdgeScooterTraversalTest.
+   */
+  @Test
+  void electricScooterUsesConstantSpeedOnHillyTerrain() {
+    // Create edge with elevation profile
+    Coordinate c1 = new Coordinate(-122.575033, 45.456773);
+    Coordinate c2 = new Coordinate(-122.576668, 45.451426);
+
+    StreetVertex from = intersectionVertex("from", c1.y, c1.x);
+    StreetVertex to = intersectionVertex("to", c2.y, c2.x);
+
+    var geometry = org.opentripplanner.framework.geometry.GeometryUtils.getGeometryFactory()
+      .createLineString(new Coordinate[] { c1, c2 });
+
+    StreetEdge hillyEdge = new StreetEdgeBuilder<>()
```

Comment:
There is a lot of duplication in this test. Can you reduce it? Not sure if it's more or less ugly but parameterized tests could also be an option.